runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java

```diff
@@ -0,0 +1,194 @@
+	private final BufferedImage tinderboxIcon;
+
+	private Color timerColor;
+	private Color warningColor;
+
+	@Inject
+	private BlastMineRockOverlay(Client client, BlastMinePlugin plugin, BlastMinePluginConfig config, ItemManager itemManager)
+	{
+		setPosition(OverlayPosition.DYNAMIC);
+		setLayer(OverlayLayer.ABOVE_SCENE);
+		this.client = client;
+		this.plugin = plugin;
+		this.config = config;
+		chiselIcon = itemManager.getImage(ItemID.CHISEL);
+		dynamiteIcon = itemManager.getImage(ItemID.DYNAMITE);
+		tinderboxIcon = itemManager.getImage(ItemID.TINDERBOX);
+	}
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		Tile[][][] tiles = client.getRegion().getTiles();
+		Widget viewport = client.getViewportWidget();
+		for (BlastMineRock rock : plugin.getRocks())
+		{
+			if (viewport != null && rock.getGameObject().getCanvasLocation() != null
+					&& rock.getGameObject().getWorldLocation().distanceTo(client.getLocalPlayer().getWorldLocation()) <= MAX_DISTANCE)
+			{
+				switch (rock.getType())
+				{
+					case NORMAL:
+						drawIconOnRock(graphics, rock, chiselIcon);
+						break;
+					case CHISELED:
+						drawIconOnRock(graphics, rock, dynamiteIcon);
+						break;
+					case LOADED:
+						drawIconOnRock(graphics, rock, tinderboxIcon);
+						break;
+					case LIT:
+						drawTimerOnRock(graphics, rock, timerColor);
+						drawAreaWarning(graphics, rock, warningColor, tiles);
+						break;
+				}
+			}
+		}
+		return null;
+	}
+
+	void updateColors()
+	{
+		timerColor = config.getTimerColor();
+		warningColor = config.getWarningColor();
+	}
+
+	private void drawIconOnRock(Graphics2D graphics, BlastMineRock rock, BufferedImage icon)
+	{
+		net.runelite.api.Point loc = Perspective.getCanvasImageLocation(client, graphics, rock.getGameObject().getLocalLocation(), icon, 150);
+		if (loc != null && config.showRockIconOverlay())
+		{
+			graphics.drawImage(icon, loc.getX(), loc.getY(), null);
+		}
+	}
+
+	private void drawTimerOnRock(Graphics2D graphics, BlastMineRock rock, Color color)
+	{
+		net.runelite.api.Point loc = Perspective.worldToCanvas(client, rock.getGameObject().getX(), rock.getGameObject().getY(), client.getPlane(), 150);
+
+		if (loc != null && config.showTimerOverlay())
+		{
+			double timeLeft = 1 - rock.getRemainingFuseTimeRelative();
+			final ProgressPieComponent pie = new ProgressPieComponent();
+			pie.setFill(color);
+			pie.setBorderColor(color);
+			pie.setPosition(loc);
+			pie.setProgress(timeLeft);
+			pie.render(graphics);
+		}
+	}
+
+	private void drawAreaWarning(Graphics2D graphics, BlastMineRock rock, Color color, Tile[][][] tiles)
+	{
+		if (!config.showWarningOverlay())
+		{
+			return;
+		}
+		int z = client.getPlane();
+		int x = rock.getGameObject().getLocalLocation().getX() / Perspective.LOCAL_TILE_SIZE;
+		int y = rock.getGameObject().getLocalLocation().getY() / Perspective.LOCAL_TILE_SIZE;
+		int orientation = tiles[z][x][y].getWallObject().getOrientationA();
```

Comment:
Mark orientation and z final and move x and y under

Reply:
Can't move orientation under since it uses x and y.

Reply:
ah, i see.